crates/bevy_ecs/src/schedule/set.rs

```diff
@@ -1,68 +1,216 @@
 use std::any::TypeId;
+use std::collections::hash_map::DefaultHasher;
 use std::fmt::Debug;
 use std::hash::{Hash, Hasher};
 use std::marker::PhantomData;
+use std::sync::RwLock;
 
 pub use bevy_ecs_macros::{ScheduleLabel, SystemSet};
-use bevy_utils::define_boxed_label;
+use bevy_utils::default;
 use bevy_utils::label::DynHash;
+use bevy_utils::once_cell::sync::OnceCell;
+use bevy_utils::HashMap;
 
 use crate::system::{
     ExclusiveSystemParamFunction, IsExclusiveFunctionSystem, IsFunctionSystem, SystemParamFunction,
 };
 
-define_boxed_label!(ScheduleLabel);
+type Interner = RwLock<HashMap<UniqueValueHash, &'static str>>;
+static INTERNER: OnceCell<Interner> = OnceCell::new();
 
-pub type BoxedSystemSet = Box<dyn SystemSet>;
-pub type BoxedScheduleLabel = Box<dyn ScheduleLabel>;
+/// The [`TypeId`](std::any::TypeId) of a value and its [`DynHash`] output when hashed with the [`DefaultHasher`].
+///
+/// This should be different for each value.
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct UniqueValueHash(TypeId, u64);
+
+/// Identifies a [`Schedule`](super::Schedule).
+pub trait ScheduleLabel: DynHash + Debug + Send + Sync + 'static {}
+
+/// A lightweight and printable identifier for a [`Schedule`](super::Schedule).
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
```

Comment:
It might be useful to implement `Hash` manually for the Id types and exclude the string literal from the hash calculation for better performance.